### ISSUE TITLE
Use Encase's Layouting Independent of Uploading

### ISSUE THREAD

**Issue by cwfitzgerald** [teoxoy/encase#16]
In [rend3's vertex pulling rewrite](https://github.com/BVE-Reborn/rend3/tree/vertex-pulling) I need to deal with sparse updates of buffers a lot. Because of that I'm not actually writing out the entire buffer all at once. I end up uploading individual items that changed into the byte offsets into the buffer they are supposed to be.

I would love to have a definition like this:

```rust
#[derive(ShaderType)]
pub struct DirectionalLightShader {
    header: DirectionalLightHeader,
    #[runtime]
    lights: Vec<DirectionalLight>,
}
```

But because I can't upload the whole thing at once, I can't actually just use WriteInto here. What I would love is an API that would give me an offset into the structure for a given member. Maybe something like

```
/// Raw Value because it's just a plain member
let offset: u64 = DirectionalLightShader::Offsets::HEADER
/// Vectors require you to give the index to get the full index
let offset: u64 = DirectionalLIghtShader::Offsets::LIGHTS::index(2);
```

This would let me use encase's layout even though I need my own custom weird layout system.

**Comment by teoxoy** [comment]
Could you point me at some code that would use this? Is this for the sparse copy compute shader you are using (where you write the new data in a new buffer and then copying it in the proper spots in the destination buffer)?

**Comment by cwfitzgerald** [comment]
Yeah that's the intended use case - using sparse copies to update a buffer that is laid out like https://github.com/BVE-Reborn/rend3/blob/trunk/rend3-routine/shaders/src/structures.wgsl#L48-L51

**Comment by teoxoy** [comment]
Actually, a variation of this feature should already be available. The `ShaderType` trait has an associated constant `METADATA` with offsets/stride.

Polishing this API would be a next step but I'd like to see how people use it first.

Let me know if you come up with something (using `METADATA`) that could provide a direction for the polished API.

cc @cwfitzgerald

**Comment by james7132** [comment]
The actual Metadata structs I think aren't documented in any form. The latest version on doc.rs doesn't even show the METADATA constant on the trait.

I'm also interested in this as we're looking into preallocating buffer space for parallelizing buffer writes in Bevy (see #https://github.com/bevyengine/bevy/pull/12489)

**Comment by teoxoy** [comment]
The metadata was meant to be internal to encase but it seems people are interested in exposing it somehow.

The main question is how would the API look like? I can see it looking quite differently depending on how it will be used.

It might be worth experimenting with an API outside encase (which would use the metadata) at first.

**Comment by stefnotch** [comment]
I also had a usecase for this. In my case, I needed the "get offset of element in an array".

I am calling [`draw_indexed_indirect`](https://docs.rs/wgpu/latest/wgpu/util/trait.RenderEncoder.html#tymethod.draw_indexed_indirect) with a buffer that contains multiple buffers for indirect drawing. 

I currently have code along the lines of
```rust
let buffer_offset = (i as u64)
                    * copy_patches::DrawIndexedIndirectArgs::METADATA
                        .alignment
                        .get();

render_pass.draw_indexed_indirect(&indirect_draw_buffers, buffer_offset);
```

It'd be nicer if I could have written something like `Vec::<copy_patches::DrawIndexedIndirectArgs>.offset_of(i)` with `offset_of` being a function that encase provides.
